### PR TITLE
Add com.obsproject.Studio.Plugin.Ocr

### PR DESCRIPTION
**Pull request by royshil**
### Please confirm your submission meets all the criteria

- [X] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [X] My pull request follows the instructions at [App Submission][submission].
- [X] I have [built][build] and tested the submission locally.
- [X] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [X] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [X] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [X] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

### CODE REVIEW

File: com.obsproject.Studio.Plugin.Ocr.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  <id>com.obsproject.Studio.Plugin.Ocr</id>
+  <extends>com.obsproject.Studio</extends>
+  <name>Ocr Plugin</name>
+  <summary>Extract and detect text in image and video inside OBS</summary>
+  <description>Using Tesseract OCR engine built-in this plugin allows detecting and extracting text from any OBS source that provide an image output and send the text to file or another source on OBS.</description>
+  <url type="homepage">https://github.com/occ-ai/obs-ocr</url>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>GPL-2.0</project_license>
+  <releases>
+    <release version="0.0.4" date="24-01-29"/>
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.metainfo.xml:12]:
```suggestion
    <release version="0.0.4" date="2024-01-29"/>
```

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+      - -DENABLE_FRONTEND_API=ON
+      - -DENABLE_QT=ON
+      - -DCMAKE_COMPILE_WARNING_AS_ERROR=ON
+      - -DCUSTOM_OPENCV_URL=/run/build/obs-ocr/opencv-linux-v4.8.1-1-Release.tar.gz
+      - -DCUSTOM_OPENCV_HASH=SHA256=809922a7cc9f344a2d82a232ed7b02e122c82e77cba94b4047e666a0527cc00e
+      - -DCUSTOM_TESSERACT_URL=/run/build/obs-ocr/tesseract-linux-5.3.3-Release.tar.gz
+      - -DCUSTOM_TESSERACT_HASH=SHA256=f4cdcb9e86374bcb9fadcf23dbe0b097622bec3fcec352b78999f90b902df0a3
+    post-install:
+      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo /run/build/obs-ocr/${FLATPAK_ID}.metainfo.xml
```
Comment (bbhtt) [inline com.obsproject.Studio.Plugin.Ocr.yaml:31]:
```suggestion
      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo ${FLATPAK_ID}.metainfo.xml
```

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        sha256: 809922a7cc9f344a2d82a232ed7b02e122c82e77cba94b4047e666a0527cc00e
+        x-checker-data:
+            type: git
+            tag-pattern: ^([-\d.]+)$
+      - type: file
+        url: https://github.com/occ-ai/obs-ocr-deps/releases/download/5.3.3-1/tesseract-linux-5.3.3-Release.tar.gz
+        sha256: f4cdcb9e86374bcb9fadcf23dbe0b097622bec3fcec352b78999f90b902df0a3
+        x-checker-data:
+          type: git
+          tag-pattern: ^v([-\d.]+)$
```
Comment (bbhtt) [inline com.obsproject.Studio.Plugin.Ocr.yaml:52]: Why are these two coming from your github org and not upstream?

Comment (bbhtt) [reply com.obsproject.Studio.Plugin.Ocr.yaml:52]: You don't need to close it. Feel free to take your time and push here.

Comment (umireon) [reply com.obsproject.Studio.Plugin.Ocr.yaml:52]: I'm working on adding a way to build dependencies from the tarballs from the upstream https://github.com/occ-ai/obs-ocr/pull/12

**Comment by umireon** [comment]
@royshil Please only re-request reviews after we are confident we meet their criteria.

**Comment by umireon** [comment]
@royshil 
> I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub.

I suppose this should be checked because you are the maintainer of obs-ocr.

File: com.obsproject.Studio.Plugin.Ocr.yaml (hunk 3 above)
Comment (bbhtt) [reply com.obsproject.Studio.Plugin.Ocr.yaml:52]: Why make it so complicated? You can addon opencv and tesseract as 2 sources in the flatpak manifest before your app.

Comment (umireon) [reply com.obsproject.Studio.Plugin.Ocr.yaml:52]: OK I will try that way.

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      - type: file
+        path: com.obsproject.Studio.Plugin.Ocr.metainfo.xml
```
[the remark this answers is not in the document]
Comment (umireon) [reply com.obsproject.Studio.Plugin.Ocr.yaml:54]: @bbhtt Do we stay in the way you expect?

[the remark this answers is not in the document]
Comment (bbhtt) [reply com.obsproject.Studio.Plugin.Ocr.yaml:54]: Yea seems fine if it works. You can link statically too if you want.

[the remark this answers is not in the document]
Comment (umireon) [reply com.obsproject.Studio.Plugin.Ocr.yaml:54]:
```suggestion
id: com.obsproject.Studio.Plugin.Ocr
branch: stable
runtime: com.obsproject.Studio
runtime-version: stable
sdk: org.kde.Sdk//6.5
build-extension: true
separate-locales: false
appstream-compose: false
build-options:
  prefix: /app/plugins/Ocr
cleanup:
  - /include
  - /share/opencv4
  - /lib/*.a
  - /lib/cmake
  - /lib/opencv4
  - /lib/pkgconfig
  - /bin
modules:
  - name: opencv
    buildsystem: cmake-ninja
    builddir: true
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
      - -DOPENCV_FORCE_3RDPARTY_BUILD=ON
      - -DBUILD_SHARED_LIBS=OFF
      - -DBUILD_opencv_apps=OFF
      - -DBUILD_opencv_js=OFF
      - -DBUILD_ANDROID_PROJECTS=OFF
      - -DBUILD_ANDROID_EXAMPLES=OFF
      - -DBUILD_DOCS=OFF
      - -DBUILD_EXAMPLES=OFF
      - -DBUILD_PACKAGE=OFF
      - -DBUILD_PERF_TESTS=OFF
      - -DBUILD_TESTS=OFF
      - -DBUILD_WITH_DEBUG_INFO=OFF
      - -DBUILD_WITH_STATIC_CRT=OFF
      - -DBUILD_WITH_DYNAMIC_IPP=OFF
      - -DBUILD_FAT_JAVA_LIB=OFF
      - -DBUILD_ANDROID_SERVICE=OFF
      - -DBUILD_CUDA_STUBS=OFF
      - -DBUILD_JAVA=OFF
      - -DBUILD_OBJC=OFF
      - -DBUILD_opencv_python3=OFF
      - -DINSTALL_CREATE_DISTRIB=OFF
      - -DINSTALL_BIN_EXAMPLES=OFF
      - -DINSTALL_C_EXAMPLES=OFF
      - -DINSTALL_PYTHON_EXAMPLES=OFF
      - -DINSTALL_ANDROID_EXAMPLES=OFF
      - -DINSTALL_TO_MANGLED_PATHS=OFF
      - -DINSTALL_TESTS=OFF
      - -DBUILD_opencv_calib3d=OFF
      - -DBUILD_opencv_core=ON
      - -DBUILD_opencv_dnn=OFF
      - -DBUILD_opencv_features2d=ON
      - -DBUILD_opencv_flann=OFF
      - -DBUILD_opencv_gapi=OFF
      - -DBUILD_opencv_highgui=OFF
      - -DBUILD_opencv_imgcodecs=OFF
      - -DBUILD_opencv_imgproc=ON
      - -DBUILD_opencv_ml=OFF
      - -DBUILD_opencv_objdetect=OFF
      - -DBUILD_opencv_photo=ON
      - -DBUILD_opencv_stitching=OFF
      - -DBUILD_opencv_video=OFF
      - -DBUILD_opencv_videoio=OFF
      - -DWITH_PNG=OFF
      - -DWITH_JPEG=OFF
      - -DWITH_TIFF=OFF
      - -DWITH_WEBP=OFF
      - -DWITH_OPENJPEG=OFF
      - -DWITH_JASPER=OFF
      - -DWITH_OPENEXR=OFF
      - -DWITH_FFMPEG=OFF
      - -DWITH_GSTREAMER=OFF
      - -DWITH_1394=OFF
      - -DWITH_PROTOBUF=OFF
      - -DBUILD_PROTOBUF=OFF
      - -DWITH_CAROTENE=OFF
      - -DWITH_EIGEN=OFF
      - -DWITH_OPENVX=OFF
      - -DWITH_CLP=OFF
      - -DWITH_DIRECTX=OFF
      - -DWITH_VA=OFF
      - -DWITH_LAPACK=OFF
      - -DWITH_QUIRC=OFF
      - -DWITH_ADE=OFF
      - -DWITH_ITT=OFF
      - -DWITH_OPENCL=OFF
      - -DWITH_IPP=OFF
    sources:
      - type: archive
        url: https://github.com/opencv/opencv/archive/refs/tags/4.9.0.tar.gz
        sha256: ddf76f9dffd322c7c3cb1f721d0887f62d747b82059342213138dc190f28bc6c
        x-checker-data:
            type: git
            tag-pattern: ^([\d.]+)$
  - name: leptonica
    buildsystem: cmake-ninja
    builddir: true
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
      - -DBUILD_SHARED_LIBS=OFF
      - -DENABLE_ZLIB=OFF
      - -DENABLE_PNG=OFF
      - -DENABLE_GIF=OFF
      - -DENABLE_JPEG=OFF
      - -DENABLE_TIFF=OFF
      - -DENABLE_WEBP=OFF
      - -DENABLE_OPENJPEG=OFF
      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
    sources:
      - type: archive
        url: https://github.com/DanBloomberg/leptonica/archive/refs/tags/1.84.1.tar.gz
        sha256: ecd7a868403b3963c4e33623595d77f2c87667e2cfdd9b370f87729192061bef
        x-checker-data:
            type: git
  - name: tesseract
    buildsystem: cmake-ninja
    builddir: true
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
      - -DBUILD_SHARED_LIBS=OFF
      - -DUSE_SYSTEM_ICU=ON
      - -DBUILD_TRAINING_TOOLS=OFF
      - -DBUILD_TESTS=OFF
      - -DDISABLED_LEGACY_ENGINE=ON
      - -DDISABLE_TIFF=ON
      - -DDISABLE_CURL=ON
      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
    sources:
      - type: archive
        url: https://github.com/tesseract-ocr/tesseract/archive/refs/tags/5.3.3.tar.gz
        sha256: dc4329f85f41191b2d813b71b528ba6047745813474e583ccce8795ff2ff5681
        x-checker-data:
            type: git
  - name: obs-ocr
    buildsystem: cmake-ninja
    builddir: true
    config-opts:
      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
      - -DENABLE_FRONTEND_API=ON
      - -DENABLE_QT=ON
      - -DCMAKE_COMPILE_WARNING_AS_ERROR=ON
      - -DUSE_SYSTEM_OPENCV=ON
      - -DUSE_SYSTEM_TESSERACT=ON
    post-install:
      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo /run/build/obs-ocr/${FLATPAK_ID}.metainfo.xml
      - appstream-compose --basename=${FLATPAK_ID} --prefix=${FLATPAK_DEST} --origin=flatpak
        ${FLATPAK_ID}
    sources:
      - type: git
        url: https://github.com/occ-ai/obs-ocr.git
        branch: umireon/system-libraries
      - type: file
        path: com.obsproject.Studio.Plugin.Ocr.metainfo.xml
```

@royshil Can you apply this suggestion?

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    config-opts:
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+      - -DENABLE_FRONTEND_API=ON
+      - -DENABLE_QT=ON
+      - -DCMAKE_COMPILE_WARNING_AS_ERROR=ON
+      - -DUSE_SYSTEM_OPENCV=ON
+      - -DUSE_SYSTEM_TESSERACT=ON
+    post-install:
+      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo /run/build/obs-ocr/${FLATPAK_ID}.metainfo.xml
```
Comment (hfiguiere) [inline com.obsproject.Studio.Plugin.Ocr.yaml:149]:
```suggestion
      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo ../${FLATPAK_ID}.metainfo.xml
```

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+sdk: org.kde.Sdk//6.5
+build-extension: true
+separate-locales: false
+appstream-compose: false
+build-options:
+  prefix: /app/plugins/Ocr
+cleanup:
+  - /include
+  - /share/opencv4
+  - /lib/*.a
```
Comment (hfiguiere) [inline com.obsproject.Studio.Plugin.Ocr.yaml:14]:
```suggestion
  - '*.a'
```

**Comment by umireon** [comment]
@bbhtt Do you expect us to make more changes?

File: com.obsproject.Studio.Plugin.Ocr.yaml (hunk 3 above)
Comment (umireon) [reply com.obsproject.Studio.Plugin.Ocr.yaml]: Implemented

File: com.obsproject.Studio.Plugin.Ocr.metainfo.xml [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  <id>com.obsproject.Studio.Plugin.Ocr</id>
+  <extends>com.obsproject.Studio</extends>
+  <name>Ocr Plugin</name>
+  <summary>Extract and detect text in image and video inside OBS</summary>
+  <description>Using Tesseract OCR engine built-in this plugin allows detecting and extracting text from any OBS source that provide an image output and send the text to file or another source on OBS.</description>
+  <url type="homepage">https://github.com/occ-ai/obs-ocr</url>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>GPL-2.0</project_license>
+  <releases>
+    <release version="0.0.4" date="2024-01-29"/>
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.metainfo.xml:12]:
```suggestion
    <release version="0.0.4" date="2024-01-29"/>
    <release version="0.0.5" date="2024-02-07"/>
```

**Comment by bbhtt** [comment]
bot, build com.obsproject.Studio.Plugin.Ocr

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+      - -DUSE_SYSTEM_OPENCV=ON
+      - -DUSE_SYSTEM_TESSERACT=ON
+    post-install:
+      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo ../${FLATPAK_ID}.metainfo.xml
+      - appstream-compose --basename=${FLATPAK_ID} --prefix=${FLATPAK_DEST} --origin=flatpak
+        ${FLATPAK_ID}
+    sources:
+      - type: git
+        url: https://github.com/occ-ai/obs-ocr.git
+        branch: umireon/system-libraries
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.yaml:155]:
```suggestion
      - type: archive
        url: https://github.com/occ-ai/obs-ocr/archive/refs/tags/0.0.5.tar.gz
        sha256: 3a04dd5774c055d0cc97520099a06b5626fe90f3fd9c3838c28d1b8466d24715
        x-checker-data:
            type: git
            tag-pattern: ^([\d.]+)$
```

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+      - -DDISABLED_LEGACY_ENGINE=ON
+      - -DDISABLE_TIFF=ON
+      - -DDISABLE_CURL=ON
+      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+    sources:
+      - type: archive
+        url: https://github.com/tesseract-ocr/tesseract/archive/refs/tags/5.3.3.tar.gz
+        sha256: dc4329f85f41191b2d813b71b528ba6047745813474e583ccce8795ff2ff5681
+        x-checker-data:
+            type: git
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.yaml:136]:
```suggestion
            type: git
            tag-pattern: ^([\d.]+)$
```

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+      - -DENABLE_TIFF=OFF
+      - -DENABLE_WEBP=OFF
+      - -DENABLE_OPENJPEG=OFF
+      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+    sources:
+      - type: archive
+        url: https://github.com/DanBloomberg/leptonica/archive/refs/tags/1.84.1.tar.gz
+        sha256: ecd7a868403b3963c4e33623595d77f2c87667e2cfdd9b370f87729192061bef
+        x-checker-data:
+            type: git
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.yaml:117]:
```suggestion
            type: git
            tag-pattern: ^([\d.]+)$
```

**Comment by umireon** [comment]
@royshil Can you add flathub.json with the following content?

```
{
    "skip-icons-check": true,
    "only-arches": ["x86_64"]
}
```

**Comment by royshil** [comment]
> @royshil Can you add flathub.json with the following content?
> 
> ```
> {
>     "skip-icons-check": true,
>     "only-arches": ["x86_64"]
> }
> ```

done

**Comment by bbhtt** [comment]
bot, build com.obsproject.Studio.Plugin.Ocr

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      - -DUSE_SYSTEM_OPENCV=ON
+      - -DUSE_SYSTEM_TESSERACT=ON
+    post-install:
+      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo ../${FLATPAK_ID}.metainfo.xml
+      - appstream-compose --basename=${FLATPAK_ID} --prefix=${FLATPAK_DEST} --origin=flatpak
+        ${FLATPAK_ID}
+    sources:
+      - type: archive
+        url: https://github.com/occ-ai/obs-ocr/releases/download/0.0.5/obs-ocr-0.0.5-source.tar.xz
+        sha256: ddf6e272e5c3a0bfa082338320b49ca146014db7e5cf8a1dc986cfdefa01e02b
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.yaml:157]:
```suggestion
        sha256: 00cc82b6828227b1a7d34b65c64de7e3b37cccd4504e15745cabcabbe0f3ab11
```

**Review by umireon — CHANGES_REQUESTED** [review]

**Comment by bbhtt** [comment]
bot, build com.obsproject.Studio.Plugin.Ocr

File: com.obsproject.Studio.Plugin.Ocr.yaml [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+      - -DCMAKE_BUILD_TYPE=RelWithDebInfo
+      - -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+      - -DENABLE_FRONTEND_API=ON
+      - -DENABLE_QT=ON
+      - -DCMAKE_COMPILE_WARNING_AS_ERROR=ON
+      - -DUSE_SYSTEM_OPENCV=ON
+      - -DUSE_SYSTEM_TESSERACT=ON
+    post-install:
+      - install -Dm644 --target-directory=${FLATPAK_DEST}/share/metainfo ../${FLATPAK_ID}.metainfo.xml
+      - appstream-compose --basename=${FLATPAK_ID} --prefix=${FLATPAK_DEST} --origin=flatpak
```
Comment (bbhtt) [inline com.obsproject.Studio.Plugin.Ocr.yaml:152]:
```
com.obsproject.Studio.Plugin.Ocr:  AppData problem: tag-invalid : <release> versions are not in order [0.0.4 before 0.0.5]
```

File: com.obsproject.Studio.Plugin.Ocr.metainfo.xml [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+    <release version="0.0.5" date="2024-02-07"/>
```
Comment (umireon) [inline com.obsproject.Studio.Plugin.Ocr.metainfo.xml:13]:
```suggestion
    <release version="0.0.5" date="2024-02-07"/>
    <release version="0.0.4" date="2024-01-29"/>
```

File: com.obsproject.Studio.Plugin.Ocr.yaml (hunk 12 above)
Comment (umireon) [reply com.obsproject.Studio.Plugin.Ocr.yaml:152]: Fixed

**Comment by bbhtt** [comment]
bot, build com.obsproject.Studio.Plugin.Ocr

**Comment by royshil** [comment]
@bbhtt @hfiguiere Anything further from your end?

**Comment by bbhtt** [comment]
No looks good.

**Pull request closed without merging**